client-spark/spark3/src/main/java/org/apache/spark/shuffle/RssShuffleManager.java

```diff
@@ -179,18 +179,20 @@ public RssShuffleManager(SparkConf conf, boolean isDriver) {
     LOG.info("Disable external shuffle service in RssShuffleManager.");
     taskToSuccessBlockIds = Maps.newConcurrentMap();
     taskToFailedBlockIds = Maps.newConcurrentMap();
-    // for non-driver executor, start a thread for sending shuffle data to shuffle server
-    LOG.info("RSS data send thread is starting");
-    eventLoop = defaultEventLoop;
-    eventLoop.start();
-    int poolSize = sparkConf.get(RssSparkConfig.RSS_CLIENT_SEND_THREAD_POOL_SIZE);
-    int keepAliveTime = sparkConf.get(RssSparkConfig.RSS_CLIENT_SEND_THREAD_POOL_KEEPALIVE);
-    threadPoolExecutor = new ThreadPoolExecutor(poolSize, poolSize * 2, keepAliveTime, TimeUnit.SECONDS,
-        Queues.newLinkedBlockingQueue(Integer.MAX_VALUE),
-        ThreadUtils.getThreadFactory("SendData-%d"));
+
     if (isDriver) {
       heartBeatScheduledExecutorService = Executors.newSingleThreadScheduledExecutor(
           ThreadUtils.getThreadFactory("rss-heartbeat-%d"));
+    } else {
+      // for non-driver executor, start a thread for sending shuffle data to shuffle server
+      LOG.info("RSS data send thread is starting");
+      eventLoop = defaultEventLoop;
+      eventLoop.start();
+      int poolSize = sparkConf.get(RssSparkConfig.RSS_CLIENT_SEND_THREAD_POOL_SIZE);
+      int keepAliveTime = sparkConf.get(RssSparkConfig.RSS_CLIENT_SEND_THREAD_POOL_KEEPALIVE);
+      threadPoolExecutor = new ThreadPoolExecutor(poolSize, poolSize * 2, keepAliveTime, TimeUnit.SECONDS,
```

Comment:
threadPoolExecutor may be null, when you stop, the Spark will throw NPE. 
Could you add a integration WordCount test case for Spark with MiniYarnCluster. like https://github.com/apache/incubator-uniffle/blob/master/integration-test/mr/src/test/java/org/apache/uniffle/test/MRIntegrationTestBase.java. Because without such integration test, some problems won't be tested.

Reply:
OK. I will add.

> threadPoolExecutor may be null, when you stop, the Spark will throw NPE.

Could u give more infos why it will be null?

Reply:
> OK. I will add.
> 
> > threadPoolExecutor may be null, when you stop, the Spark will throw NPE.
> 
> Could u give more infos why it will be null?

You only assign a value to the threadPoolExecutor when the process is not driver. So the threadPoolExecutor will be null when the process is driver.

Reply:
But the threadPool is only invoked by the executor and closed when it's null.

Reply:
Driver will call method `stop`,too. You can see https://github.com/Tencent/Firestorm/pull/187